### PR TITLE
ACI: Move dev_guide from Wiki to Ansible docs

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
We had this item on our list for some time, as the Wiki is not the right location for documentation.
https://github.com/ansible/community/wiki/Network:-ACI-Development-Guide

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
aci dev_guide

##### ANSIBLE VERSION
v2.7

##### ADDITIONAL INFORMATION
Screenshot of local Table of Contents:
--------
screenshot from 2018-09-13 05-53-27
--------

### CODE REVIEW

**Comment by dagwieers** [comment]
cc @jmcgill298

**Review by acozine — CHANGES_REQUESTED** [review]
Thanks for contributing content @dagwieers. This PR needs some work. At a minimum, please make Shippable happy, make the headers conform to https://github.com/ansible/ansible/issues/35520, add a basic description (what is an ACI module and why would someone want to write their own?), and add a link to this page from the ACI guide. Active headings would make this page better as well. When that's all done, I'll give the text a more thorough read.

**Comment by dagwieers** [comment]
@mattclay @gundalow @acozine The rstcheck sanity check is being more strict than Sphinx itself. It is complaining about python examples (even snippets of code) whereas Sphinx doesn't care. Which means I now have to ensure that my snippets of code are syntactically correct (which is silly, because we used place-holders to make code more clear or we selected pieces of code instead of full code).

**Comment by dagwieers** [comment]
@acozine Can you please merge #35520 so a contributor can simply look at the existing documentation to ensure he's using the correct conventions ? We cannot expect contributors to have an active knowledge of open PR's and their authority (as long as something is not merged, IMO it cannot hold authority over what is already merged...)

**Comment by dagwieers** [comment]
@acozine What are "Active headings" ?

**Comment by mattclay** [comment]
The syntax checking can be an inconvenience for some examples. Fortunately those cases can be easily resolved with minor syntax changes or a small amount of boilerplate. The alternative would be to disable syntax checking, but that seems like giving up more than we'd be gaining.

**Comment by dagwieers** [comment]
@mattclay Ah, so this is a known problem. I wonder if there's a way to disable the syntax checking for those examples we know this is not a problem.

We explain the module's structure, by using snippets from a module. Which means we break up `if elif else` statements into separate examples. There's no easy way to fix that unless we break that structure and then it does no longer reflect the module.

**Comment by acozine** [comment]
@dagwieers thanks for the quick response - what time is it where you are???

#35520 is an Issue, not a PR, so I can't merge it, but if you look at the recent changes to the Developer Guide (#45179) or the Community Guide (#45364) you will see the header notation in action. I'm working to fix the headers slowly, along with other changes, which is why the issue is still open. If you have specific questions, about header notation, let me know.

**Comment by acozine** [comment]
By "active headings" I meant giving headings verbs and more detail - for example, "Importing shared ACI code" instead of "Imports". You can ignore that for now - the other changes are more important.

**Comment by dagwieers** [comment]
@acozine It's 4AM over here. If you need any help fixing the remaining documentation to follow the new rules, let me know. I now based my headings on the Windows dev_guide, because I am not sure what are supposed to be parts and chapters in my document.

PS While improving the docs I was doing exactly that!

**Comment by acozine** [comment]
Some of the headings in this doc are already active: "Testing for sanity checks" is a good example.

**Comment by mattclay** [comment]
@dagwieers For the if/elif/else case, perhaps something like the following would suffice:

```python
    if conditional:
        pass  # other logic here
    else
        # actual example here
```

Yes, this adds boilerplate, but it also passes the tests. Also, for some readers it will be easier to follow since the example is free standing. Otherwise, when starting reading in the middle of the document it can look like part of the example is missing.

**Comment by dagwieers** [comment]
@mattclay It adds confusion because it no longer matches with reality. You can no longer add the snippets together to get the complete module. So I think in this case trying to accommodate the syntax checking is making the documentation worse. And that's the world upside down, the syntax checking should help us write better/flawless documentation, and not make us change examples (for the worse) because it breaks the (in)sanity checks :-) That's why a way to selectively disable it would be a better fit IMO.

**Comment by mattclay** [comment]
It's definitely a matter of personal preference. Different readers will prefer one approach vs the other.

@acozine What direction would you like to take with the docs?

**Comment by dagwieers** [comment]
@mattclay I don't think any reader would appreciate stub-code to be added only to accommodate the syntax checker. In fact, they won't know stuff was added because of a syntax-checker. They shouldn't be aware of any of this...

**Comment by dagwieers** [comment]
And, to repeat, the content is perfectly acceptable by Sphinx (and GitHub's parser) and the examples are even highlighted fine.

**Comment by mattclay** [comment]
I find having the "stub code" at the top of the conditional useful for providing context. But that's just me. I'll leave it up to you and @acozine to work out what you'd like to do.

**Comment by dagwieers** [comment]
Well, thinking about this even more, I happily give up syntax highlighting if that keeps the examples identical to the modules.

**Comment by dagwieers** [comment]
@acozine I think I fixed all your remarks. Not all headings are *active* because I don't think it makes sense everywhere. Feedback appreciated.

**Comment by dagwieers** [comment]
@acozine Since this is a PR from 2 different authors (the original document and then the changes) I would prefer if both commits are merged using "Rebase and merge" rather than "Squash and merge" so the content is being correctly attributed.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst:21]: Add a link to the ACI object model here? I found https://www.cisco.com/c/en/us/td/docs/switches/datacenter/aci/apic/sw/1-x/aci-fundamentals/b_ACI-Fundamentals/b_ACI-Fundamentals_chapter_010001.html - not sure if that is the right thing.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Hyphenate "module-specific parameters" (x2).

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Currently present/absent/query are in _italics_ - did you mean them to be in `code`?

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: I'd change the hint to "Do not provide default values for configuration arguments. Default values could cause unintended changes to the object."

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Instead of "check_mode", I'd either enclose it in double backticks to make it look like `code`, or say "check mode" without the underscore. Also, currently state is _italic_ (single backtick), should probably be `code` (double backticks).

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: "its child configuration" not "it's child confguration"
Also, either put "str" in double backticks or spell out "string"

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Can you make this note active voice? Instead of "If no ID parameters are supplied", say "if you do not supply any ID parameters" or "if your task does not supply any ID parameters" . . . and so on for the rest of the paragraph.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: I can't tell if "vlanns" is a typo or not (two Ns).

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst:202]: Active voice here if possible - "Once you've built the URL and filter string, your module is . . ."

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: I'd put double backticks around `present` here.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: "is done" is a bit confusing here - maybe "The configuration s built the same way as the object" or "The configuration portion is built just like the object portion"?

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Use sentence case for headings (When state is present, not When State is Present).

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [inline docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Sentence case here too.

File: docs/docsite/rst/scenario_guides/guide_aci.rst
Comment (acozine) [inline docs/docsite/rst/scenario_guides/guide_aci.rst:336]: Thanks for finding and fixing this typo!

**Review by acozine — CHANGES_REQUESTED** [review]
Overall this looks great. Thanks @dagwieers for including anchors for incoming links, as well as for making Shippable happy and fixing the headers. I appreciate your attention to detail. I've found a few little things to fix before merging.

**Comment by acozine** [comment]
For anyone else following this PR, the changes are temporarily visible at http://docs.testing.ansible.com/ansible/latest/dev_guide/developing_modules_general_aci.html.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst:21]: Well, I contemplated using some of the drawings, because they are really helpful, but didn't want to go to the trouble getting the rights to reuse them. But a link to the document could be useful. Fact is that I consider people planning to write such modules to know about this already. Anyone responsible for working with Cisco ACI is already trained and possibly even certified as this is not something small shops will be using.

Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst:21]: I'll add some links, it doesn't hurt.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: The original docs were written by someone else, I guess so.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Done.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Done.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Done.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Done.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Is that a general theme in documentation ? I usually try to avoid directing the user where I can.

Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: In this case the person we are instructing is writing the module, other people are using the module. That may be the reason the text was impersonal. I don't know. I changed it nevertheless.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: It's correct, it means "vlan namespace".

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst:202]: It's not the reader that is building the URL and filter string, it is the code that does.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Done.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Done.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Done.

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (dagwieers) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst]: Done.

**Comment by dagwieers** [comment]
@acozine @jmcgill298 Please re-review !

File: docs/docsite/rst/dev_guide/developing_modules_general_aci.rst
Comment (acozine) [reply docs/docsite/rst/dev_guide/developing_modules_general_aci.rst:21]: ah, I was thinking of an inline link, but the "see also" section works too

**Review by acozine — APPROVED** [review]

**Comment by acozine** [comment]
Looks good on the writing side, thanks @dagwieers. I'll let @jmcgill298 weigh in before merging.

**Comment by jmcgill298** [comment]
LGTM

**Comment by dagwieers** [comment]
@acozine Can we backport this to v2.7 ?

**Comment by acozine** [comment]
@dagwieers go ahead and open a backport PR, if we can't get it in before the release, I'll merge it the day after and republish the docs.

**Pull request merged**
